content/posts/Gitops.md

```diff
@@ -0,0 +1,35 @@
+
+Using Ian’s Tree we see that GitOps is a continuation of the story that started out with shell scripts, Makefiles, Dockerfiles, Jenkinsfiles, building reusable images with Packer, infrastructure-as-code tools like Terraform, and configuration management tools such as Ansible, Puppet, and Chef. GitOps brings some new tools to this ecosystem for managing containerized applications in Kubernetes clusters.
+
+Weaveworks (the hardworking folks behind [Flux](https://github.com/fluxcd/flux) and [eksctl](https://github.com/weaveworks/eksctl)) have been talking about GitOps for some time now, and have a useful explanation of the term [here](https://www.weave.works/technologies/gitops/). 
+
+The idea is to manage applications declaratively (something that should sound familiar when talking about Kubernetes manifests) and automatically applying changes using code that is stored in a central repository (`git` being one of the more popular source control tools in use these days).
+
+Imagine having a central git repository that contains all of your Kubernetes manifests, both for your own applications, along with any supporting services you need to install, such as [prometheus](https://prometheus.io/) for metrics, [fluentd](https://www.fluentd.org/) for logs, or [cert-manager](https://github.com/jetstack/cert-manager) for managing TLS certificates. This gives you a single place to look to see who deployed what and when, all tracked in the git commit history.
+
+Adding a new application to your clusters could be as simple as opening a pull-request to this central “configuration repo” with your Kubernetes manifests or Helm Charts for your new service. Once that PR is approved by your team and merged, a GitOps tool (like Flux or [Argo](https://argoproj.github.io/argo-cd/) could automatically apply that change from within the cluster.
+
+If this sounds similar to a Puppet or Chef agent automatically installing a package across your fleet of Linux VMs when you push the change into your config repo, then you can see how the past is very much a part of the present.
+
+Another popular workflow that GitOps tools can provide is to automatically deploy upgrades when a new container image is pushed to your container image registry. Imagine a testing cluster where your GitOps tool could automatically deploy new container images of your app as soon as they are built and published to the registry, rather than having a developer do a separate deployment to promote the new image. The GitOps tool would automatically make a commit back to the central configuration repo when the deploy is finished setting the newly applied container tag.
+
+If you needed to rebuild your entire Kubernetes cluster from scratch now you have everything you need in a single repo to recreate everything. Simply install your GitOps tool, point it at your central configuration repo, and it will automatically reinstall everything using the last deployed versions.
```

Comment:
```suggestion
A common problem with Kubernetes is how to recreate the 'state of the cluster'. It's not that you actually often _need_ to rebuild a cluster from scratch, but the point is that in a declarative infrastructure there should be an automatable _way_ to do that. Otherwise, recovering from disasters is a complex and error-prone manual process which may rely on information you no longer have available.

 The centralized config repo and the GitOps tooling offers a way to deal with that. Simply install your GitOps tool, point it at your repo, and it will automatically reinstall everything using the last deployed versions, bringing the cluster back to a state where it's ready for production. It's rare to be able to make this _entirely_ automated, because there is usually also some data in the system, such as your user database, or uploaded assets, for example. But using the GitOps discipline certainly helps with everything that _can_ be managed declaratively, leaving you more time to deal with the stuff that needs human intervention.
```